src/main/scala/com/netflix/edda/Crawler.scala

```diff
@@ -135,12 +147,68 @@ abstract class Crawler extends Observable {
           if (logger.isDebugEnabled) logger.debug(s"$req$this sending: $msg -> $o")
           o ! msg
       })
+      /* reset the error count at the end of each run */
+      retry_count = 0
       setLocalState(state, CrawlerState(records = newRecords))
 
       // } else state
     }
   }
 
+  /** wrapper for all requests to throttle or delay access to the AWS API */
+  def backoffRequest[T](code: =>T): T = {
+    // using a ratio of 2:1 for errors to retries due to requests being sent concurrently
+    val errorReducer = 2
+    // number of "free" errors to ignore before full throttling is enabled
+    val margin = 10
+    val marginalDelay = ( throttle_delay.get.toInt * (((retry_count-margin)/errorReducer) + 1) )
+    val throttleDelay = ( throttle_delay.get.toInt * ((retry_count/errorReducer) + 1) )
+
+    /* configurable delay for all requests to slow down access */
+    if (request_delay.get.toInt > 0) Thread sleep request_delay.get.toInt
+
+    /* configurable random delay to further throttle access */
+    if (jitter.get.toBoolean) {
+      val maxJitter = Utils.getProperty("edda.crawler", "jitter.max", name, "2000").get.toInt
+      val rand = new Random
+      val jitter = (maxJitter * rand.nextDouble).toLong
+      logger.debug("{} api request delayed by {}ms", this, jitter)
+      Thread sleep jitter
+    }
+
+    /* handler for AWS imposed throttling */
+    if (throttle.get.toBoolean) {
+      try {
+        if (retry_count > margin) {
+          logger.debug("{} SLEEPING [{}]: {}", Array(this, retry_count, marginalDelay))
+          Thread sleep marginalDelay
+        }
+        code
+      } catch {
+        case e: AmazonClientException => {
+          val pattern = ".*Error Code: ([A-Za-z]+);.*".r
+          val pattern(err_code) = e.getMessage()
+          if ( (err_code == "RequestLimitExceeded") || (err_code == "Throttling") ) {
+            Thread sleep throttleDelay
+            retry_count = retry_count + 1
+            if ((retry_count/errorReducer) >= retry_max.get.toInt) {
+              logger.error("Hit configured maximum number of API backoff requests, aborting")
+              throw e
+            }
+            backoffRequest { code }
+          } else {
+            logger.error("Unexpected AmazonClientException, aborting")
```

Comment:
Since the exception is propagated it will likely be logged later on, but it might be useful to include the exception in this error log so it is easy for someone to see what the unexpected exception was.